cairo/ethereum/cancun/transactions.cairo

```diff
@@ -215,3 +224,166 @@ func signing_hash_4844{range_check_ptr, bitwise_ptr: BitwiseBuiltin*, keccak_ptr
     let hash = keccak256(encoded_tx);
     return hash;
 }
+
+func recover_sender{
+    range_check_ptr,
+    range_check96_ptr: felt*,
+    add_mod_ptr: ModBuiltin*,
+    mul_mod_ptr: ModBuiltin*,
+    bitwise_ptr: BitwiseBuiltin*,
+    keccak_ptr: KeccakBuiltin*,
+    poseidon_ptr: PoseidonBuiltin*,
+}(chain_id: U64, tx: Transaction) -> (Address, EthereumException*) {
+    alloc_locals;
+    tempvar SECP256K1N = U256(new U256Struct(low=secp256k1.N_LOW_128, high=secp256k1.N_HIGH_128));
+    tempvar SECP256K1N_DIVIDED_BY_2 = U256(
+        new U256Struct(low=secp256k1.N_DIVIDED_BY_2_LOW_128, high=secp256k1.N_DIVIDED_BY_2_HIGH_128)
+    );
+    tempvar zero = U256(new U256Struct(low=0, high=0));
+
+    if (cast(tx.value.legacy_transaction.value, felt) != 0) {
+        let r = tx.value.legacy_transaction.value.r;
+        let s = tx.value.legacy_transaction.value.s;
+
+        let r_is_zero = U256__eq__(r, zero);
+        let r_is_out_of_range = U256_le(SECP256K1N, r);
+
+        let s_is_zero = U256__eq__(s, zero);
+        let s_is_within_range = U256_le(s, SECP256K1N_DIVIDED_BY_2);
+
+        let error = r_is_zero.value + r_is_out_of_range.value + s_is_zero.value + (
+            1 - s_is_within_range.value
+        );
+        if (error != 0) {
+            tempvar err = new EthereumException(InvalidSignatureError);
+            return (Address(0), err);
```

Comment:
considering that the error is never caught and acted upon, we can assert 1 = 0 and wrap in a `with_attr` (for all errors in this function).

Reply:
and change the signature to return only the address

Reply:
i tried it like this but it didn't work in the runner / serde for strict_raises

Reply:
since in unit tests, python will raise the ethereumexception in python, and in cairo an assert_eq

Reply:
done